### DISCUSSION TITLE
Custom useQuery/useMutation hooks for TRPC

### DISCUSSION THREAD

**Discussion by Eliav2** [trpc/trpc#4638] [Q&A]
Currently, when using the `useQuery` and `useMutation` hooks provided by the popular `react-query` library, developers often wrap these hooks with custom hooks to add additional logic and error handling. For instance, a common practice is to create a custom hook like `useAppQuery`, which internally uses `useQuery` but also includes custom error handling or other specific behavior.

**for Example:**

```jsx
// useAppQuery.js
import { useQuery } from 'react-query';

export function useAppQuery(queryKey, queryFn, options) {
  const result = useQuery(queryKey, queryFn, options);

  // Custom error handling or additional logic can be added here
  if (result.error) {
    // Handle the error, show a toast, etc.
  }

  return result;
}
```

if could be great TRPC could provide a way to use a custom hook such 'useAppQuery' instead of the normal 'useQuery'.
_same applies for useMutation/useQueries_
 
However, when working with the client-side TRPC's `useQuery` and `useMutation`, it appears that there is no straightforward way to provide similar custom hooks. This makes it challenging for developers to implement reusable logic, error handling, or additional customization for these TRPC hooks.

so currently in order to wrap 'users' query hook for example:
```jsx
// YourComponent.js
import { trpc } from 'trpc'; // Assuming trpc is imported from the client-side library

function YourComponent() {
  // Currently, there is no simple way to create a custom hook like useAppQuery for trpc.<procedure>.useQuery
  const { data, error } = trpc.users.useQuery('yourQuery', { variables: { ... } });

  // Component logic here
}
```
could be achieved with something like this:
```jsx
export function useCustomUserQuery(input, options) {
  const { data, error } = trpc.users.useQuery(input, options);

  // Custom error handling or additional logic can be added here
  if (error) {
    // Handle the error, show a toast, etc.
  }

  return { data, error };
}
```
However, if you want to customize other TRPC hooks, like trpc.posts.useQuery or trpc.comments.useQuery, you would need to create separate custom hooks for each of them. not very convenient. 

The feature could enhance the TRPC client-side library by providing a mechanism similar to `react-query` that allows developers to create custom hooks for `useQuery` and `useMutation`. This would enable easier implementation of reusable logic and error handling, improving code organization and maintainability for applications using TRPC.

Maybe this is already possible today? could not find anything about it. Any help would be appreciated!

**Comment by Eliav2** [comment]
currently using this wrapper on each query works fine:
```jsx
const useTrpcQuery = <T extends UseTRPCQueryResult<any, any>>(
  trpcQueryhook: () => T
): T => {
  const query = trpcQueryhook();
  if (query.error) {
    console.log(query.error);
  }

  return query;
};

function App() {
  const greetings = useTrpcQuery(() => trpc.greeting.useQuery({ name }));
  greetings.data // string as expected

  // component logic 
}

```

complete file:
```jsx
// useTrpcHooks.tsx
import {
  UseTRPCMutationResult,
  UseTRPCQueryResult,
} from "@trpc/react-query/shared";

export const useTrpcQuery = <T extends UseTRPCQueryResult<any, any>>(
  trpcQueryHook: () => T
): T => {
  const query = trpcQueryHook();
  if (query.error) {
    console.log(query.error);
  }

  return query;
};

export const useTrpcMutation = <
  T extends UseTRPCMutationResult<any, any, any, any>
>(
  trpcMutationHook: () => T
): T => {
  const mutation = trpcMutationHook();

  const trpcMutation = mutation.mutate;
  mutation.mutate = async (input) => {
    trpcMutation(input);
  };

  if (mutation.error) {
    console.log(mutation.error);
  }

  return mutation as any;
};

```

if you have better approach, feel free to share.

**Comment by craxrev** [comment]
In my case I had to find a way where I don't have to rewrite all my `useQuery` / `useMutation` calls, so I had to "redo" the trpc proxy object by parsing its props recursively and refactor `useQuery` / `useMutation` methods:
```ts
const createWrapperProxy = (value: any) => {
  return new Proxy(value, {
    get: (target, prop, receiver): any => {
      const propValue = Reflect.get(target, prop, receiver);

      if (typeof prop !== "string") return undefined;

      if (prop === "useQuery") {
        return (input: any, opts?: any) => {
          const useQuery = propValue;

          // Do stuff here

          const query = useQuery(input, {
            ...opts,
            onSuccess: (...args: any) => {
              opts?.onSuccess?.(...args);
            },
            onError: (...args: any) => {
              opts?.onError?.(...args);
              // Or here
            },
          });

          return query;
        };
      } else if (prop === "useMutation") {
        return (opts?: any) => {
          const useMutation = propValue;

          // Do stuff here

          const mutation = useMutation({
            ...opts,
            onSuccess: (...args: any) => {
              opts?.onSuccess?.(...args);
            },
            onError: (...args: any) => {
              opts?.onError?.(...args);
              // Or here
            },
          });

          return mutation;
        };
      } else {
        return createWrapperProxy(propValue);
      }
    },
  });
};

const api: typeof trpc = createWrapperProxy(trpc);
```
